### DISCUSSION TITLE
Kotlin coroutines

### DISCUSSION THREAD

**Discussion by tomvandenberge** [spring-projects/spring-integration#3894] [Q&A]
Hi,
I'm using Spring Integration in various Kotlin (previously Java) projects. Some of these projects are using coroutines. So far, I haven't been able to find a way to use coroutines together with Spring Integration. I'd love to be able e.g. to have a suspending function as a service activator. Currently this is not possible.

I realise that this probably requires fundamental changes to Spring Integration, but I wanted to check with you if this is something that is already on your roadmap?

Thanks,
Tom

**Comment by artembilan** [comment]
I'm not familiar with coroutines, so give me, please, some time to learn what is that.
We do support, though, `CompletableFuture` and `org.reactivestreams.Publisher` return types, so probably this `suspend` can be implemented some way over there as well.
As well as a `@MessagingGateway` may benefit from this Kotlin feature:

https://docs.spring.io/spring-integration/docs/current/reference/html/messaging-endpoints.html#async-service-activator
https://docs.spring.io/spring-integration/docs/current/reference/html/reactive-streams.html#reactive-reply-payload
https://docs.spring.io/spring-integration/docs/current/reference/html/messaging-endpoints.html#async-gateway

**Reply by artembilan** [reply]
BTW [contribution](https://github.com/spring-projects/spring-integration/blob/main/CONTRIBUTING.adoc) is welcome!

**Reply by tomvandenberge** [reply]
Thanks, that's very much appreciated.

How exactly coroutines might work together with Spring Integration is not totally clear to me, as it is a vast area to cover.  However the first (overly) simple use case that comes to my mind is a situation where I send a message to a message gateway, which is then sent to a service activator, and all of this in a single thread. Both the code to send the message to the gateway and the service activator are Kotlin suspending functions. If the Integration components in between would also somehow be suspending functions, the entire flow could be executed as a coroutine. This allows the executing thread to be released when the coroutine is suspended anywhere.

I'll certainly provide a contribution when I can think of something useful.
